powerstore/resource_snapshot.go

```diff
@@ -0,0 +1,270 @@
+package powerstore
+
+import (
+	"context"
+	"fmt"
+	"github.com/dell/gopowerstore"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"log"
+	"regexp"
+	"terraform-provider-powerstore/client"
+	"terraform-provider-powerstore/models"
+)
+
+// newSnapshotResource returns snapshot new resource instance
+func newSnapshotResource() resource.Resource {
+	return &resourceSnapshot{}
+}
+
+type resourceSnapshot struct {
+	client *client.Client
+}
+
+// Metadata defines resource interface Metadata method
+func (r *resourceSnapshot) Metadata(ctx context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_snapshot"
+}
+
+// Schema defines resource interface Schema method
+func (r *resourceSnapshot) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+
+		MarkdownDescription: "volume snapshot resource",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed:            true,
+				Description:         "The unique identifier of the volume snapshot.",
+				MarkdownDescription: "The unique identifier of the volume snapshot.",
+			},
+			"name": schema.StringAttribute{
+				Required:            true,
+				Description:         "Name of the volume snapshot.",
+				MarkdownDescription: "Name of the volume snapshot.",
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+			"volume_id": schema.StringAttribute{
+				Optional:            true,
+				Description:         "ID of the volume to take snapshot.",
+				MarkdownDescription: "ID of the volume to take snapshot.",
+				Validators: []validator.String{
+					stringvalidator.LengthAtLeast(1),
+				},
+			},
+			"description": schema.StringAttribute{
+				Optional:            true,
+				Computed:            true,
+				Description:         "Description of the volume snapshot.",
+				MarkdownDescription: "Description of the volume snapshot.",
+			},
+			"performance_policy_id": schema.StringAttribute{
+				Optional:            true,
+				Computed:            true,
+				Description:         "Performance Policy id of the volume snapshot.",
+				MarkdownDescription: "Performance Policy id of the volume snapshot.",
+				Validators: []validator.String{
+					stringvalidator.OneOf([]string{
+						"default_medium",
+						"default_low",
+						"default_high",
+					}...),
+				},
+			},
+			"expiration_timestamp": schema.StringAttribute{
```

Comment:
Can we support desired retention and retention unit params so that it will be convenient for the user ?